deep_field_metadetect/jaxify/jax_metacal.py

```diff
@@ -0,0 +1,636 @@
+            weight_fac=0.5,
+            max_min_fft_size=max_min_fft_size,
+        )
+
+    # Use vmap to parallelize across shears
+    vectorized_shear_op = jax.vmap(single_shear_op)
+    mcal_obs_list = vectorized_shear_op(g1_vals, g2_vals)
+
+    # Convert back to dictionary format
+    mcal_res = {}
+    for i, shear in enumerate(shears):
+        mcal_res[shear] = jax.tree.map(lambda x: x[i], mcal_obs_list)
+
+    return mcal_res
+
+
+@partial(
+    jax.jit,
+    static_argnames=[
+        "nxy",
+        "nxy_psf",
+        "return_k_info",
+        "force_stepk_field",
+        "force_maxk_field",
+        "force_stepk_psf",
+        "force_maxk_psf",
+        "max_min_fft_size",
+    ],
+)
+def jax_match_psf(
+    dfmd_obs,
+    reconv_psf,
+    nxy,
+    nxy_psf,
+    return_k_info=False,
+    force_stepk_field=0.0,
+    force_maxk_field=0.0,
+    force_stepk_psf=0.0,
+    force_maxk_psf=0.0,
+    max_min_fft_size=1024,
+):
+    """Match the PSF on an dfmd observation to a new PSF."""
+    wcs = dfmd_obs.wcs._local_wcs
+    image = get_jax_galsim_object_from_dfmd_obs(
+        dfmd_obs,
+        kind="image",
+        force_stepk=force_stepk_field,
+        force_maxk=force_maxk_field,
+    )
+    psf = get_jax_galsim_object_from_dfmd_obs(
+        dfmd_obs.psf,
+        kind="image",
+        force_stepk=force_stepk_psf,
+        force_maxk=force_maxk_psf,
+    )
+
+    ims = jax_galsim.Convolve(
+        [image, jax_galsim.Deconvolve(psf), reconv_psf],
+        gsparams=jax_galsim.GSParams(
+            minimum_fft_size=max_min_fft_size,
+            maximum_fft_size=max_min_fft_size,
+        ),
+    )
+
+    ims = ims.withGSParams(
+        minimum_fft_size=max_min_fft_size,
+        maximum_fft_size=max_min_fft_size,
+    )
+    ims = ims.drawImage(nx=nxy, ny=nxy, wcs=wcs).array
+
+    def return_obs_and_kinfo(_):
+        return _jax_render_psf_and_build_obs(
+            ims, dfmd_obs, reconv_psf, nxy_psf, weight_fac=1
+        ), (image.stepk, image.maxk, psf.stepk, psf.maxk)
+
+    def return_obs_only(_):
+        return _jax_render_psf_and_build_obs(
+            ims, dfmd_obs, reconv_psf, nxy_psf, weight_fac=1
+        ), (0.0, 0.0, 0.0, 0.0)
+
+    return jax.lax.cond(
+        return_k_info, return_obs_and_kinfo, return_obs_only, operand=None
+    )
+
+
+def _extract_attr(obs, attr, dtype=jnp.float64):
+    if getattr(obs, "has_" + attr)():
+        return getattr(obs, attr)
+    else:
+        return np.zeros_like(obs.image, dtype=dtype)
+
+
+@partial(jax.jit, static_argnames=["ignore_psf", "skip_mfrac_for_second"])
+def jax_add_dfmd_obs(
+    dfmd_obs1, dfmd_obs2, ignore_psf=False, skip_mfrac_for_second=False
+) -> DFMdetObservation:
+    """Add two dfmd observations"""
+
+    if repr(dfmd_obs1.wcs) != repr(dfmd_obs2.wcs):
+        raise RuntimeError(
+            "AffineTransforms must be equal to add dfmd observations! %s != %s"
+            % (repr(dfmd_obs1.wcs), repr(dfmd_obs2.wcs)),
+        )
```

Comment:
This test likely does nothing when running the code.

Reply:
added a comment